src/mapping/query/functions.rs

```diff
@@ -0,0 +1,444 @@
+use super::Function;
+use crate::{
+    event::{Event, Value},
+    mapping::Result,
+    types::Conversion,
+};
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ToStringFn {
+    query: Box<dyn Function>,
+    default: Option<Value>,
+}
+
+impl ToStringFn {
+    pub(in crate::mapping) fn new(query: Box<dyn Function>, default: Option<Value>) -> Self {
+        Self { query, default }
+    }
+}
+
+impl Function for ToStringFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let result = match self.query.execute(ctx) {
+            Ok(v) => match v {
+                Value::Bytes(_) => Ok(v),
+                _ => Ok(Value::Bytes(v.as_bytes())),
+            },
+            Err(err) => Err(err),
+        };
+        if result.is_err() {
+            if let Some(v) = &self.default {
+                return Ok(v.clone());
+            }
+        }
+        result
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ToIntegerFn {
+    query: Box<dyn Function>,
+    default: Option<Value>,
+}
+
+impl ToIntegerFn {
+    pub(in crate::mapping) fn new(query: Box<dyn Function>, default: Option<Value>) -> Self {
+        Self { query, default }
+    }
+}
+
+impl Function for ToIntegerFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let result = match self.query.execute(ctx) {
+            Ok(v) => match v {
+                Value::Integer(_) => Ok(v),
+                Value::Float(f) => Ok(Value::Integer(f as i64)),
+                Value::Bytes(_) => Conversion::Integer.convert(v).map_err(|e| e.to_string()),
+                Value::Boolean(b) => Ok(Value::Integer(if b { 1 } else { 0 })),
+                Value::Timestamp(t) => Ok(Value::Integer(t.timestamp())),
+                Value::Null => Err("value is null".to_string()),
+                _ => Err("unable to convert array or object into int".to_string()),
+            },
+            Err(err) => Err(err),
+        };
+        if result.is_err() {
+            if let Some(v) = &self.default {
+                return Ok(v.clone());
+            }
+        }
+        result
```

Comment:
```suggestion
        match self.query.execute(ctx) {
            Ok(v) => match v {
                Value::Integer(_) => Ok(v),
                Value::Float(f) => Ok(Value::Integer(f as i64)),
                Value::Bytes(_) => Conversion::Integer.convert(v).map_err(|e| e.to_string()),
                Value::Boolean(b) => Ok(Value::Integer(if b { 1 } else { 0 })),
                Value::Timestamp(t) => Ok(Value::Integer(t.timestamp())),
                Value::Null => Err("value is null".to_string()),
                _ => Err("unable to convert array or object into int".to_string()),
            },
            Err(err) => Err(err),
        }.or_else(|err| match &self.default {
            Some(v) => Ok(v.clone()),
            None => Err(err),
        })
```

Ditto for the other `if result.is_err()` below.

Since that `or_else` clause is repeated above, it might be worth a helper, not sure.